Miscellaneous trait look-up reductions

reaperrr opened the pull request:
Some more trait look-up reductions in places that might not get called that often, but potentially several times at once.

I was too lazy to measure the performance impact (unlikely to be noteworthy, anyway), but those felt simple enough to pass as "can't hurt and might add up, without regressing or over-complicating code".

reaperrr committed:
Save Mobile trait look-up in AttackBase

reaperrr committed:
Check for IsReloading in HasAnyValidWeapons

For Attack*.CanAttack().

Allows us to drop the additional check for armaments with ammo.

reaperrr committed:
Use Mobile.Pathfinder to reduce trait look-ups

anvilvapre commented on OpenRA.Mods.Common/Activities/Move/MoveAdjacentTo.cs:
```diff
@@ -45,8 +44,9 @@ public MoveAdjacentTo(Actor self, in Target target, WPos? initialTargetPosition
-			pathFinder = self.World.WorldActor.Trait<IPathFinder>();
+
+			// PERF: Because we can be sure that OccupiesSpace is Mobile here, we can save some performance by avoiding querying for the trait.
+			Mobile = (Mobile)self.OccupiesSpace;
```
Mods won't be able to add multiple Mobile traits to an actor? In the past it would cause an exception. Now it would cause OccupiesSpace to be used.

reaperrr replied:
Not sure what you mean? Actors are only allowed to have one of these traits, and Actor.OccupiesSpace is just `IOccupySpace` cached by the actor, which is implemented by Mobile, so in a way we're already caching Mobile and can use this casting to save a look-up.

anvilvapre replied:
That was my question. Can actors only have one of these traits.

pchote replied:
Actors can only have one of these traits currently, and many things will need to be reworked before that can change.

anvilvapre reviewed (commented):
Mods won't be able to add multiple Mobile traits to an actor? In the past it would cause an exception during trait lookup. Now it would cause OccupiesSpace to be used?

reaperrr commented:
Updated
Removed first commit less because of the comment, but more because attempted profiling showed MoveAdjacentTo to be far less common than I thought.

Didn't bother profiling the `Mobile.Pathfinder` changes since that's more polish than anything.

Perf improvements from the 2 `CanAttack` changes, from left to right (bleed, armament changes only, both), in stopwatch ticks:
![canatk-perf](https://user-images.githubusercontent.com/2857877/108599994-a70e0500-7394-11eb-971c-e9c6ea0fdeff.png)

pchote commented on OpenRA.Mods.Common/Traits/Attack/AttackBase.cs:
```diff
@@ -152,7 +152,8 @@ protected virtual bool CanAttack(Actor self, in Target target)
 				return false;
 
-			var mobile = self.TraitOrDefault<Mobile>();
+			// PERF: Mobile implements IFacing, so we can use 'as' to save a trait look-up here.
```
Minor style nit: IMO it feels more natural to use `IPositionable` instead of `IFacing` here, but since it doesn't really matter for practical purposes I don't consider this to be a blocker.

pchote commented on OpenRA.Mods.Common/Traits/Attack/AttackBase.cs:
```diff
@@ -149,17 +149,14 @@ protected virtual bool CanAttack(Actor self, in Target target)
 				return false;
 
-			if (!HasAnyValidWeapons(target))
+			if (!HasAnyValidWeapons(target, checkForReloadingState: true))
```
Style nit: IMO its not immediately clear what `checkForReloadingState` means, without reading the implementation - specifically the "state" in the name makes me think that this might be doing something more complicated than it really is. How about something like `requireLoaded` or `reloadingIsInvalid`?

pchote commented on OpenRA.Mods.Common/Traits/Attack/AttackBase.cs:
```diff
@@ -241,7 +238,8 @@ public bool HasAnyValidWeapons(in Target t, bool checkForCenterTargetingWeapons
 			{
 				var checkIsValid = checkForCenterTargetingWeapons ? armament.Weapon.TargetActorCenter : !armament.IsTraitPaused;
-				if (checkIsValid && !armament.IsTraitDisabled && armament.Weapon.IsValidAgainst(t, self.World, self))
+				var reloadingStateIsValid = checkForReloadingState ? !armament.IsReloading : true;
```
This can simplify to `!checkForReloadingState || !armament.IsReloading`, and i'd again drop the "state" from the variable name.

pchote reviewed (commented):
Looks reasonable to me, just had a couple of small style nits.

reaperrr commented:
Updated.

abcdefg30 reviewed (approved):
Code changes lgtm but untested

pchote commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/2259475ff1e3ce0f818821bb82eb208b0fea8960
